src/overlays/actors/ovl_Bg_Mori_Rakkatenjo/z_bg_mori_rakkatenjo.c

```diff
@@ -20,35 +42,199 @@ const ActorInit Bg_Mori_Rakkatenjo_InitVars = {
+void BgMoriRakkatenjo_Init(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad;
+    BgMoriRakkatenjo* this = THIS;
+    ColHeader* colHeader = NULL;
+
+    DynaPolyInfo_SetActorMove(&this->dyna, 1);
+    // Forest Temple obj. Falling Ceiling
+    osSyncPrintf("森の神殿 obj. 落下天井 (home posY %f)\n", this->dyna.actor.initPosRot.pos.y);
+    if ((fabsf(1991.0f - this->dyna.actor.initPosRot.pos.x) > 0.001f) ||
+        (fabsf(683.0f - this->dyna.actor.initPosRot.pos.y) > 0.001f) ||
+        (fabsf(-2520.0f - this->dyna.actor.initPosRot.pos.z) > 0.001f)) {
+        // The set position has been changed. Let's fix the program.
+        osSyncPrintf("Warning : セット位置が変更されています。プログラムを修正しましょう。\n");
+    }
+    if (this->dyna.actor.initPosRot.rot.y != 0x8000) {
+        // The set Angle has changed. Let's fix the program.
+        osSyncPrintf("Warning : セット Angle が変更されています。プログラムを修正しましょう。\n");
+    }
+    this->moriTexObjIndex = Object_GetIndex(&globalCtx->objectCtx, OBJECT_MORI_TEX);
+    if (this->moriTexObjIndex < 0) {
+        // Forest Temple obj Falling Ceiling Bank Danger!
+        osSyncPrintf("Error : 森の神殿 obj 落下天井 バンク危険！(%s %d)\n", "../z_bg_mori_rakkatenjo.c", 205);
+        Actor_Kill(&this->dyna.actor);
+        return;
+    }
+    Actor_ProcessInitChain(&this->dyna.actor, sInitChain);
+    DynaPolyInfo_Alloc(&D_060087AC, &colHeader);
+    this->dyna.dynaPolyId =
+        DynaPolyInfo_RegisterActor(globalCtx, &globalCtx->colCtx.dyna, &this->dyna.actor, colHeader);
+    BgMoriRakkatenjo_SetupWaitForMoriTex(this);
+    sCamSetting = 0;
+}
+
+void BgMoriRakkatenjo_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad;
+    BgMoriRakkatenjo* this = THIS;
+
+    DynaPolyInfo_Free(globalCtx, &globalCtx->colCtx.dyna, this->dyna.dynaPolyId);
+}
+
+s32 BgMoriRakkatenjo_LinkUnder(BgMoriRakkatenjo* this, GlobalContext* globalCtx) {
+    Vec3f* pos = &PLAYER->actor.posRot.pos;
+
+    return (-3300.0f < pos->z) && (pos->z < -1840.0f) && (1791.0f < pos->x) && (pos->x < 2191.0f);
+}
+
+s32 BgMoriRakkatenjo_LinkClose(BgMoriRakkatenjo* this, GlobalContext* globalCtx) {
+    Vec3f* pos = &PLAYER->actor.posRot.pos;
+
+    return (-3360.0f < pos->z) && (pos->z < -1840.0f) && (1791.0f < pos->x) && (pos->x < 2191.0f);
+}
+
+void BgMoriRakkatenjo_SetupWaitForMoriTex(BgMoriRakkatenjo* this) {
+    this->actionFunc = BgMoriRakkatenjo_WaitForMoriTex;
+}
+
+void BgMoriRakkatenjo_WaitForMoriTex(BgMoriRakkatenjo* this, GlobalContext* globalCtx) {
+    if (Object_IsLoaded(&globalCtx->objectCtx, this->moriTexObjIndex)) {
+        BgMoriRakkatenjo_SetupWait(this);
+        this->dyna.actor.draw = func_808A5D60;
+    }
+}
+
+void BgMoriRakkatenjo_SetupWait(BgMoriRakkatenjo* this) {
+    this->timer = (this->fallCount > 0) ? 100 : 21;
+    this->dyna.actor.posRot.pos.y = 683.0f;
+    this->actionFunc = BgMoriRakkatenjo_Wait;
+}
+
+void BgMoriRakkatenjo_Wait(BgMoriRakkatenjo* this, GlobalContext* globalCtx) {
```

Comment:
for consistency with `BgMoriRakkatenjo_Rest` and clarity, what about renaming
`BgMoriRakkatenjo_Wait` to `BgMoriRakkatenjo_RestCeiling`/High/Top
`BgMoriRakkatenjo_Rest` to `BgMoriRakkatenjo_RestFloor`/Low/Bottom

Reply:
I prefer the different names. In "Wait", the ceiling is waiting for something to happen (Link moving under it), while in "Rest", it's just pausing for a bit on the ground before it starts moving upward again.

Reply:
Makes sense. For some reason I thought it only waited for Link's position on the first fall :)